### PR TITLE
Parse Manifest Descriptor fields

### PR DESCRIPTION
**Pull request by aakropotkin**
This covers parsing for manifest _descriptors_, being entries under `installs.<ID> = <DESCRIPTOR>`.

### CODE REVIEW

File: include/flox/resolver/descriptor.hh [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    std::optional<std::string> name;
+
+    /** Match `version` or `semver` if a modifier is present. */
+    std::optional<std::string> version;
+
+    /** Match a catalog stability. */
+    std::optional<std::string> stability;
+
+    /** @brief A dot separated attribut path, or list representation. */
+    using Path = std::variant<std::string, flox::AttrPath>;
```
Comment (mkenigs) [inline include/flox/resolver/descriptor.hh:51]: Should we settle on a single format for these? Would this expose either format to users?

File: src/resolver/descriptor.cc [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+          this->input =
+            parseFlakeRef( std::get<std::string>( * raw.packageRepository ) );
+        }
+      else
+        {
+          this->input = nix::FlakeRef::fromAttrs(
+            std::get<nix::fetchers::Attrs>( * raw.packageRepository )
+          );
+        }
+      assert( std::holds_alternative<nix::FlakeRef>( * this->input ) );
```
Comment (mkenigs) [inline src/resolver/descriptor.cc:299]: Could `input` just be a `FlakeRef`?

File: include/flox/resolver/descriptor.hh [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+                                               )
+
+
+/* -------------------------------------------------------------------------- */
+
+/**
+ * @brief Extend and remap fields from @a flox::resolver::PkgDescriptorRaw to
+ *        those found in a `flox` _manifest_.
+ */
+struct ManifestDescriptor {
```
Comment (mkenigs) [inline include/flox/resolver/descriptor.hh:115]: I'm not sure if this class is providing much value. `fillPkgQueryArgs` doesn't have much logic - it seems like it's mostly just copying values. Could we go straight from `ManifestDescriptorRaw` to `PkgQueryArgs`?

File: src/resolver/descriptor.cc [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+  const auto & second = glob.at( 1 );
+  if ( second.has_value() &&
+       ( ( * second ) != "null" ) && ( ( * second ) != "*" )
+     )
+    {
+      desc.systems = std::vector<std::string> { * second };
+      if ( raw.systems.has_value() && ( * raw.systems != * desc.systems ) )
+        {
+          throw FloxException(
+            "`systems' list conflicts with `absPath' system specification"
```
Comment (mkenigs) [inline src/resolver/descriptor.cc:211]: There's a few places you're checking if values conflict. It seems like code/logic might be simpler to just forbid simultaneously setting any fields that could conflict. And I almost think that would be less confusing to use. Why would a user want to repeat themself?

File: include/flox/resolver/descriptor.hh [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    void clear();
+
+    /**
+     * @brief Fill a @a flox::pkgdb::PkgQueryArgs struct with preferences to
+     *        lookup packages.
+     *
+     * NOTE: This DOES NOT clear @a pqa before filling it.
+     * This is intended to be used after filling @a pqa with global preferences.
+     * @param pqa A set of query args to _fill_ with preferences.
+     * @return A reference to the modified query args.
```
Comment (mkenigs) [inline include/flox/resolver/descriptor.hh:165]: Thought: I find it hard to understand exactly how `fillPkgQueryArgs` fits in with the rest of the codebase, because it's not clear what the state of the `PkgQueryArgs` is before it gets modified here, and it's not clear what modifications might happen afterwards. And since there are so many different definitions of `fillPkgQueryArgs` it's a bit hard to understand how they all fit together. I wonder if at some point it might be more clear to split things up into something like query args + registry args + preferences args, with each one of those having a simpler lifecycle, and then there could be one place they were all pulled together into `PkgQueryArgs`

File: include/versions.hh [hunk 6]
```diff
@@ -45,12 +47,33 @@ bool isDate( const std::string & version );
 /** @return `true` iff @a version can be interpreted as _semantic version_. */
 bool isCoercibleToSemver( const std::string & version );
 
+/**
+ * @brief Determine if @a version is a valid _semantic version range_ string.
+ *
+ * This is far from a complete check, but it should be sufficient for our usage.
```
Comment (mkenigs) [inline include/versions.hh:53]: What do we need to be correct for our usage? Would there be any options for checking this without implementing it ourselves?

File: tests/manifest.cc [hunk 7]
```diff
@@ -57,19 +61,371 @@ test_yamlToJSON0()
...
+  /* We expect this to be recognized as an _indirect flake reference_. */
+  EXPECT( descriptor.input.has_value() );
+  EXPECT( std::holds_alternative<nix::FlakeRef>( * descriptor.input ) );
+  auto flakeRef = std::get<nix::FlakeRef>( * descriptor.input );
+
+  EXPECT_EQ( flakeRef.input.getType(), "indirect" );
+
+  auto alias = flakeRef.input.attrs.at( "id" );
+  EXPECT( std::holds_alternative<std::string>( alias ) );
+  EXPECT_EQ( std::get<std::string>( alias ), "nixpkgs" );
```
Comment (mkenigs) [inline tests/manifest.cc:102]: Seems like a few of these tests would be cleaner if you defined `operator==` on `ManifestDescriptor`

File: include/flox/core/util.hh [hunk 8]
```diff
@@ -351,6 +351,28 @@ bool isUInt( std::string_view str );
...
+
+
+/** @brief trim from start ( copying ). */
+[[nodiscard]] std::string ltrim_copy( std::string_view str );
+
+/** @brief trim from end ( copying ). */
+[[nodiscard]] std::string rtrim_copy( std::string_view str );
+
+/** @brief trim from both ends ( copying ). */
+[[nodiscard]] std::string trim_copy( std::string_view str );
```
Comment (mkenigs) [inline include/flox/core/util.hh:373]: Should we use boost instead?

**Review by mkenigs — CHANGES_REQUESTED** [review]

File: include/flox/resolver/descriptor.hh (hunk 1 above)
Comment (aakropotkin) [reply include/flox/resolver/descriptor.hh:51]: I'd prefer a list, but the spec indicates that either a list of strings or a "dot separated path" may be used.

File: include/flox/resolver/descriptor.hh [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+     */
+    using AbsPath = std::variant<std::string, AttrPathGlob>;
+    /** Match an absolute path, allowing globs for `system`. */
+    std::optional<AbsPath> absPath;
+
+    /** Only resolve for a given set of systems. */
+    std::optional<std::vector<std::string>> systems;
+
+    /** Whether resoution is allowed to fail without producing errors. */
+    std::optional<bool> optional;
```
[the remark this answers is not in the document]
Comment (aakropotkin) [reply include/flox/resolver/descriptor.hh:67]: Yes they're both useful.

File: include/flox/resolver/descriptor.hh [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+    /** Named _group_ that the package is a member of. */
+    std::optional<std::string> packageGroup;
+
+    /** Force resolution is a given input or _flake reference_. */
+      std::optional<std::variant<std::string, nix::fetchers::Attrs>>
+    packageRepository;
+
+    /** Relative path to a `nix` expression file to be evaluated. */
+    std::optional<std::string> input;
```
[the remark this answers is not in the document]
Comment (aakropotkin) [reply include/flox/resolver/descriptor.hh:77]: I really don't want to do that one any time soon; but to align with the spec I've added it to the parser.

File: include/flox/resolver/descriptor.hh [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    std::optional<AbsPath> absPath;
+
+    /** Only resolve for a given set of systems. */
+    std::optional<std::vector<std::string>> systems;
+
+    /** Whether resoution is allowed to fail without producing errors. */
+    std::optional<bool> optional;
+
+    /** Named _group_ that the package is a member of. */
+    std::optional<std::string> packageGroup;
```
[the remark this answers is not in the document]
Comment (aakropotkin) [reply include/flox/resolver/descriptor.hh:70]: The spec indicates that groups are tagged with a `group: "blue"` label, so this follows the spec.
Once we have a working PoC we can hash out any refinement on the schema.

For now lets get a first iteration that works <3

File: include/flox/resolver/descriptor.hh [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+                                               , packageRepository
+                                               , input
+                                               )
+
+
+/* -------------------------------------------------------------------------- */
+
+/**
+ * @brief Extend and remap fields from @a flox::resolver::PkgDescriptorRaw to
+ *        those found in a `flox` _manifest_.
```
[the remark this answers is not in the document]
Comment (aakropotkin) [reply include/flox/resolver/descriptor.hh:113]: They're similar but not exactly the same.

This handles renamed fields and junk. I imagine we can flatten/align these in the near future though.

File: src/resolver/descriptor.cc (hunk 2 above)
Comment (aakropotkin) [reply src/resolver/descriptor.cc:299]: No because it could also be a path to a nix expression file ( it can't yet, but from the perspective of the interface it is expected to in the future ).

File: include/flox/resolver/descriptor.hh (hunk 3 above)
Comment (aakropotkin) [reply include/flox/resolver/descriptor.hh:115]: This class holds the code which converts from one to the other.
Think of it as more of a closure/functor that transforms `ManifestDescriptorRaw -> PkgQueryArgs`.

It doesn't have to be a class all, it could just be a function `manifestDescriptorToQuery` or something with a bunch of helper functions; but for the purposes of testing, debugging, and organizing the docs a class felt ergonomic to me.

File: include/flox/core/util.hh (hunk 8 above)
Comment (aakropotkin) [reply include/flox/core/util.hh:373]: We could, I just dropped these in to avoid digging around boost and adding it as a direct dependency.
This works for now.

File: include/versions.hh (hunk 6 above)
Comment (aakropotkin) [reply include/versions.hh:53]: We could tweak `node-semver` to expose additional subcommands for parts of its parsers.

Similar to what we've done with `parser-util` and `flox`.

File: include/flox/resolver/descriptor.hh (hunk 11 above)
[the remark this answers is not in the document]
Comment (zmitchell) [reply include/flox/resolver/descriptor.hh:70]: We already have a spec, I don't think this PR is the place to talk about more design decisions.

File: include/flox/resolver/descriptor.hh (hunk 9 above)
[the remark this answers is not in the document]
Comment (mkenigs) [reply include/flox/resolver/descriptor.hh:67]: My recollection was that `optional` was originally used to allow a package to only work on some systems. When else would a user use it?

File: include/flox/resolver/descriptor.hh (hunk 12 above)
[the remark this answers is not in the document]
Comment (mkenigs) [reply include/flox/resolver/descriptor.hh:113]: I understand `PkgDescriptor` as mapping `PkgDescriptorRaw` to `PkgQueryArgs`, not `PkgDescriptorRaw` to manifest, so I don't really understand the comment

File: include/versions.hh (hunk 6 above)
Comment (mkenigs) [reply include/versions.hh:53]: If you're going to leave as is, can you explain why it's sufficient for our usage? I.e. if we need to change it, what kinds of things need to work? And what kinds of things can be incomplete?

File: src/resolver/descriptor.cc [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+        /* If it's a valid semver or a date then it's not a range. */
+        if ( versions::isSemver( trimmed ) || versions::isDate( trimmed ) ||
+             ( ! versions::isSemverRange( trimmed ) )
+           )
+          {
+            desc.version = std::move( trimmed );
+          }
+        else /* Otherwise, assume a range. */
+          {
+            desc.semver = std::move( trimmed );
```
[the remark this answers is not in the document]
Comment (mkenigs) [reply src/resolver/descriptor.cc:73]: Non-blocking - curious if this understanding is correct

File: src/resolver/descriptor.cc [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+    {
+      return std::get<AttrPathGlob>( absPath );
+    }
+  AttrPathGlob   glob;
+  flox::AttrPath path = splitAttrPath( std::get<std::string>( absPath ) );
+  size_t         idx  = 0;
+  for ( const auto & part : path )
+    {
+      /* Treat `null' or `*' in the second element as a glob. */
+      if ( ( idx == 1 ) && ( ( part == "null" ) || ( part == "*" ) ) )
```
[the remark this answers is not in the document]
Comment (mkenigs) [reply src/resolver/descriptor.cc:96]: Blocking if we're not throwing when we should

File: src/resolver/descriptor.cc [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+        {
+          const auto & elem = * itr;
+          if ( ! elem.has_value() )
+            {
+              throw FloxException(
+                "`absPath' may only have a glob as its second element"
+              );
+            }
+          desc.path->emplace_back( * elem );
+        }
```
[the remark this answers is not in the document]
Comment (mkenigs) [reply src/resolver/descriptor.cc:184]: Nonblocking

File: tests/manifest.cc (hunk 7 above)
Comment (mkenigs) [reply tests/manifest.cc:102]: Nonblocking

File: include/flox/resolver/descriptor.hh (hunk 1 above)
Comment (mkenigs) [reply include/flox/resolver/descriptor.hh:51]: I don't get the sense from the spec that's deliberate - if it is deliberate, great. If not, I think we should tweak it

File: include/flox/resolver/descriptor.hh (hunk 9 above)
[the remark this answers is not in the document]
Comment (aakropotkin) [reply include/flox/resolver/descriptor.hh:67]: Context from the spec ( paraphrased ):

> From the users' perspective `optional` and `systems` could be used to solve similar categories of problems; but `systems` being given an explicit field allows us to optimize resolution by quite a bit, and is more ergonomic for users who "want this for Darwin but not Linux".
> 
> `optional` is more general and simply says "if resolution/build fails, skip/_warn and ignore_".

Having said that, lets avoid discussions about things we like/don't like about the spec here. They're good conversations but moving them to a product ticket and covering them after a first iteration will help our velocity.

File: include/flox/resolver/descriptor.hh (hunk 1 above)
Comment (aakropotkin) [reply include/flox/resolver/descriptor.hh:51]: @graham made this tweak when I bumped into this a few weeks back so I think we're good.

File: include/versions.hh (hunk 6 above)
Comment (aakropotkin) [reply include/versions.hh:53]: It's sufficient for a first iteration.
It checks the 99% case to distinguish between a range and "static" version.

The main reason to detect this is because from the CLI we can't immediately tell whether `<NAME>@<VERSION-OR-SEMVER>` is an exact version match ( like a date ), or a real range.

This does a "best effort" detection which is fine for now in my opinion.

File: include/flox/resolver/descriptor.hh (hunk 12 above)
[the remark this answers is not in the document]
Comment (aakropotkin) [reply include/flox/resolver/descriptor.hh:113]: It just means "We take the base concept of a descriptor and add some extra fields. also some fields in the manifest have different names than those used by the `flox search` JSON queries, so we handle those naming differences here"

File: include/versions.hh (hunk 6 above)
Comment (mkenigs) [reply include/versions.hh:53]: Cool can you add a comment saying it's checking the 99% case which is _, and not the 1% case which is _? I'm just wanting to understand so that if we have to touch it without you we'll have some idea of what needs to work

File: src/resolver/descriptor.cc [hunk 16]
```diff
@@ -0,0 +1,382 @@
...
+    {
+      initManifestDescriptorAbsPath( * this, raw );
+    }
+  else if ( raw.stability.has_value() )
+    {
+      this->subtree   = Subtree( "catalog" );
+      this->stability = raw.stability;
+    }
+
+  /* Only set if it wasn't handled by `absPath`. */
```
[the remark this answers is not in the document]
Comment (aakropotkin) [reply src/resolver/descriptor.cc:245]: it refers to the earlier block

File: src/resolver/descriptor.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (aakropotkin) [reply src/resolver/descriptor.cc:73]: This is a good question and it was one I needed to talk with Graham and y'all about.

I realized that from the perspective of the CLI's `flox install <NAME>@<VERSION-OR-SEMVER-RANGE>` invocation it's ambiguous which is being indicated ( for example `@4.0` may mean "4.0.*" ( range ) or "=4.0" ( exact string match ).
To account for that ambiguity I knew we needed to distinguish the difference in the manifest as well when we add _short_ inline descriptors like `installs.foo = foo@4.2.0`.
So I just made it a single field in the exploded form to and didn't align precisely with the spec.

I wanted to talk that over with everyone and potentially make a tweak but when I was doing dev and testing I just had the "hard" case so I knew I was really testing it.
Separating back out to align with the spec is a small change though.

This was definitely a good one to flag.

File: src/resolver/descriptor.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (aakropotkin) [reply src/resolver/descriptor.cc:96]: IDK I have a bunch of error checks all over the place in here and a ton of tests so I think it's likely caught.
You can add more tests in a follow up though or recommend patches to improve the logic if you'd like.

File: include/versions.hh (hunk 6 above)
Comment (aakropotkin) [reply include/versions.hh:53]: I expanded the inline comment

Comment (mkenigs) [reply include/versions.hh:53]: Thanks!

File: src/resolver/descriptor.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (aakropotkin) [reply src/resolver/descriptor.cc:96]: I have a struct definition like that in `github:flox/resolver` actually.
It just never made the migration because it felt like _another class_ that I only really needed in like 1 routine so I just in-lined it.

https://github.com/flox/resolver/blob/main/include/flox/types.hh#L115

Comment (mkenigs) [inline src/resolver/descriptor.cc:96]:
```suggestion
      /* Treat `null' or `*' as a glob. */
      /* TODO we verify that only the second option is a glob elsewhere, but we could do that here instead */
      if ( ( ( part == "null" ) || ( part == "*" ) ) )
```

File: include/flox/resolver/descriptor.hh (hunk 12 above)
Comment (mkenigs) [inline include/flox/resolver/descriptor.hh:113]:
```suggestion
 * @brief Extend and remap fields from @a flox::resolver::PkgDescriptorRaw to
 *        @a flox::pkgdb::PkgQueryArgs
```

File: src/resolver/descriptor.cc (hunk 14 above)
Comment (aakropotkin) [reply src/resolver/descriptor.cc:96]: Since we've already done it in the block above I'm not sure that it makes sense to repeat those checks here

Comment (mkenigs) [reply src/resolver/descriptor.cc:96]: What block are these checks done in?

Comment (aakropotkin) [reply src/resolver/descriptor.cc:96]: https://github.com/flox/pkgdb/pull/76/files#diff-c7d6d0eb80d5d13a6a8424de32efea1281bfd24fefc950e83dedb274d6cec038R144

Comment (mkenigs) [reply src/resolver/descriptor.cc:96]: That line only checks for nullopt, not "null" or "*", which should be done here. If you tweak the manifest test like this:
```
diff --git a/tests/manifest.cc b/tests/manifest.cc
index 706c314..86a61f4 100644
--- a/tests/manifest.cc
+++ b/tests/manifest.cc
@@ -252,7 +252,7 @@ test_parseManifestDescriptor_path0()
 {

   flox::resolver::ManifestDescriptorRaw raw = R"( {
-    "absPath": "legacyPackages.null.hello"
+    "absPath": "legacyPackages.null.null"
   } )"_json;

   flox::resolver::ManifestDescriptor descriptor( raw );
@@ -262,7 +262,7 @@ test_parseManifestDescriptor_path0()
   EXPECT( ! descriptor.systems.has_value() );
   EXPECT( ! descriptor.stability.has_value() );
   EXPECT( descriptor.path.has_value() );
-  EXPECT( ( * descriptor.path ) == ( flox::AttrPath { "hello" } ) );
+  EXPECT( ( * descriptor.path ) == ( flox::AttrPath { "null" } ) );

   return true;
 }
```
It will currently pass. With my patch, it will fail with:
```
Testing: tests/manifest
  ERROR: parseManifestDescriptor_path0: `absPath' may only have a glob as its second element
```

Comment (aakropotkin) [reply src/resolver/descriptor.cc:96]: I see what you mean. I was confused about "checking" to mean "exception handling".
I think you're talking about the parsing?

That's a good catch. `["legacyPackages",null,"null"]` for a package named `null`.

Comment (mkenigs) [reply src/resolver/descriptor.cc:96]: By the TODO I mean it might make more sense to verify that only the second element is a glob here, rather than higher up in multiple places. But whatever you prefer 🤷 

Regardless, "legacyPackages.null.null" and "legacyPackages.null.*" need to ERROR, and currently they aren't.

**Review by mkenigs — APPROVED** [review]

File: include/flox/resolver/descriptor.hh (hunk 1 above)
Comment (ghudgins) [reply include/flox/resolver/descriptor.hh:51]: it was deliberate at the time but I wasn't thorough when I PRd it. it snuck in on another. if we could go back to the spec just calling for `.` then i'm good. if there's a "some people will prefer lists and it's near zero cost to implement both" then i'm good. even if we had to do lists it's not the end of the world because this way of expressing packages will likely be something advanced users do. 

I can put it back to dots since it seems like we now support both.

File: src/resolver/descriptor.cc (hunk 14 above)
Comment (aakropotkin) [reply src/resolver/descriptor.cc]: There is technically a package named `legacyPackages.null.null` by our spec which is what I thought you were highlighting.

There's a package named `null`. Your test case shows that I didn't properly handle that edge case, so that's definitely something to fix in a follow up PR.

Thanks for highlighting the gap :v:
